### ISSUE TITLE
[SECURITY ISSUE] DotNet.nugetPush leaks API key to log

### ISSUE THREAD

**Issue by zvirja** [fsharp/FAKE#2526]
### Description

Current implementation of `dotnet nuget push` wrapper is not using `TraceSecrets` to protect API key from leaking.
It a security issue, as people could accidentally leak their key in CI logs. Probably, there are already projects affected by this, as I found a lot of projects using this util.

### Repro steps

Use API as following:

```c#
DotNet.nugetPush (fun p -> { p with PushParams = { p.PushParams with ApiKey = Some apiKey }})
                 pkg
```

### Expected behavior

ApiKey is not logged

### Actual behavior

ApiKey is logged:

```
C:\Project> "C:\Program Files\dotnet\dotnet.EXE"  nuget push "C:\Project\package.nupkg" --api-key dummy_API_KEY --source https://feed (In: false, Out: false, Err: false)
```

### Known workarounds

~~Use `NuGet.NuGet.NuGetPublish` instead, which is not ideal, as it requires `nuget.exe` tool presence.~~
**UPD:** Issue is reproducible for NuGetPublish as well.

### Related information 

* Version of FAKE: Fake.DotNet.NuGet 5.20.1

**Comment by zvirja** [comment]
@matthid Just found that no matter how you publish packages, key is leaked. I think it's a critical security issue and build logs on CI might already contain keys.

**Comment by matthid** [comment]
Can you manually call `TraceSecrets.register` to protect the key?

> critical security issue

Note that I wouldn't recommend building your security on top of this feature, this is a very light fence which any developer can circumvent (the same is true for the corresponding feature on Azure DevOps https://github.com/microsoft/azure-pipelines-tasks/issues/4284)

**Comment by zvirja** [comment]
> Can you manually call TraceSecrets.register to protect the key?

@matthid Thanks! Yes, was just about to update the issue and mention this workaround. It works.

> Note that I wouldn't recommend building your security on top of this feature,

Well, it depends what you call a security issue. Open source projects are using this feature to automate their package publication, like I'm doing for [AutoFixture](https://github.com/AutoFixture/AutoFixture). So leakage of their NuGet key might be a bit concerning.

I'm not sure I got what exactly you recommend - do you advise to not use this feature on CI at all? AppVeyor CI I'm using is already providing facilities to protect the key. They don't decrypt values for PRs, so it's not possible to dump the value for strangers. OTOH if code is merged, then key is not protected anymore. I would expect the key be never leaked when FAKE is used in legitimate way. Is that expectation not correct?

I'm about to create a PR and fix those two findings.

**Comment by matthid** [comment]
> Well, it depends what you call a security issue. Open source projects are using this feature to automate their package publication, like I'm doing for AutoFixture. So leakage of their NuGet key might be a bit concerning.

Yes, I use the same here for FAKE. What I meant is:
- You shouldn't blindly assume all FAKE APIs call `TraceSecrets.register` internally. (Or that FAKE completely prevents everything). PR is welcome, but I'm pretty sure a lot of APIs don't do this "automagically".
- You should make sure to test first before depending on this feature (including the failing case).
- Understand the security implications (for example it doesn't protect you when you do a `printfn "secret"`)

So in a nutshell just don't think FAKE will magically protect you against any leakage. You should combine this with the native CI capability, and see this as a fallback for locally running the script and test both separately and combined.

>  I would expect the key be never leaked when FAKE is used in a legitimate way. Is that expectation not correct? 🤔
> I'm about to create a PR and fix those two findings.

Yes, that would be the expectation in a perfect world, but we don't have the resources here to promise something like that :) I hope the explanation above makes sense.

Thanks for taking care of this and sending PRs!

**Comment by zvirja** [comment]
@matthid I see your point, thanks for the explanation. Interestingly, that your statement about perfect world applies to the points above in a reverse direction - don't expect people to manually test the feature and not just blindly assume that all it happens "automagically" 😅  Personally, I found it by accident, as I believed that library already handled it. Therefore, I would say that as a library author I would say never assume that clients are responsible enough 😉 Luckily, MIT license protects us as authors from liability, so at least that is out of concerns.

I like that in FAKE 5 you made a step forward and make the `TracingSecrets` feature global. So it's just enough to register it in one place and _theoretically_ all other places are protected. Previously, as far as I remember protecting the secret was really tricky and annoying, as it required ad-hoc actions all the time. Nice work on this!
